### PR TITLE
fix(ioslides): Shiny-specific initialization

### LINKED ISSUE #2514: shiny presentation renders empty black screen
Hi. I ran the default template for a Shiny Presentation and it rendered an empty, black screen with no errors. All of my packages are updated and I'm running the latest versions of RStudio Desktop and R. I've attached the code I ran and the versions I'm using in two separate text files.  I appreciate your help in advance. 

[r_version.txt](https://github.com/rstudio/rmarkdown/files/12577726/r_version.txt)
[demo_template.txt](https://github.com/rstudio/rmarkdown/files/12577772/demo_template.txt)

### PR DESCRIPTION
**Pull request by gadenbuie**
Fixes #2514 
Fixes rstudio/shiny#3894

Currently, ioslides waits for the `DOMContentLoaded` event to initialize the slides, assuming that the `<slides>` markup will be there when that event happens.

Beacuse the slides are loaded via dynamic UI with `uiOutput()` (not just the slides, but the app content for any Shiny app in R Markdown), and due to recent changes in Shiny, the dynamic UI isn't present in the DOM yet. This is due to recent changes in Shiny that make dynamic UI rendered asynchronous.

This PR is a minimal fix that adds special initialization code for Shiny in the ioslides pandoc template and prevents the ioslides `slide-deck.js` from trying to initialize slides automatically in a shiny context.

This works because the slides markup, including the new initialization code, are all added to the page at the same time, so by the time `loadDeck()` is called we have a strong guarantee that `<slides>` is present.

cc @cderv @cpsievert (I can't request reviewers)

### CODE REVIEW

File: inst/rmd/ioslides/ioslides-13.5.1/js/slide-deck.js
Comment (cderv) [inline inst/rmd/ioslides/ioslides-13.5.1/js/slide-deck.js:805]: Here we are deferring if no `slides` which happens with new async behavior 

So we could also adapt that part to correctly wait for this element, right ?

**Review by cderv — COMMENTED** [review]
Ok good. this is a simple enough solution. That suppose to change the external deps `slide-deck.js` to add shiny specific thing, in two places. So maybe I'll move the loading in the template. 

However, it seems the issue is that `<slides>` element is not there yet. So my first thought would have been to improve the defer code until the element exists. 

Would that work too ? 

I'll also wait for @cpsievert (or anyone else from shiny team) to review before merging.

File: inst/rmd/ioslides/ioslides-13.5.1/js/slide-deck.js
Comment (cderv) [reply inst/rmd/ioslides/ioslides-13.5.1/js/slide-deck.js:805]: It seems when Shiny is on the page is the problem here. `slide-deck.js` is run before `windows.Shiny` exists it seems. 

So we could do what you say by moving the whole `loadDeck()` call outside of `slide-deck.js` to be called inside the pandoc template that will be inserted as UI. 

I don't know if there is a way to make `slide-deck.js` loads after Shiny.

Comment (gadenbuie) [reply inst/rmd/ioslides/ioslides-13.5.1/js/slide-deck.js:805]: It's not worth the complexity. Technically, you could have been calling `loadDeck()` in the ioslides pandoc template this whole time. But I'd still keep this code here for portability reasons.

Similarly, yes there's lots of refactoring that could happen in this code but I don't recommend making any changes right now.

@cpsievert The Shiny task queue isn't a good fit here for a few of reasons. But primarily this file doesn't use jQuery and I don't think it should use Shiny methods either, beyond just testing for its presence.

Comment (cderv) [reply inst/rmd/ioslides/ioslides-13.5.1/js/slide-deck.js:805]: Ok great ! 

Thanks for your feedback. All good - I'll accept this change.

**Review by cderv — APPROVED** [review]

**Pull request merged**
